sql: use new foreign key representation

jordanlewis opened the pull request:
This PR does the following salient things:

- change all in-memory foreign key operations to use the new foreign key
  representation, which is defined on table descriptor instead of index
  descriptor.
- change all reads of serialized table descriptor protos to
  unconditionally upgrade themselves from the old descriptor version
  into the new descriptor version.
- change all serialization paths of table descriptors to downgrade
  themselves from the new descriptor version to the old descriptor
  version *when the cluster version is less than 19.2*.

Note that this commit *is not intended to change* any actual foreign key
behavior, such as restrictions on dropping or flexibility in adding.

Note further that this commit puts the state of this branch into peril
with regards to backup and restore on mixed version clusters, which
won't work. This is a known issue and will be resolved before 19.2 is
released. See issue #39474.

There are a couple of other relevant technical details.

New-style foreign keys contain 4 special fields that are designed to
help with this upgrade.

Two are the "pinned" indexes that the foreign key is related to. These
will be the same as the indexes that the foreign keys used to live on in
19.1 and before, if the foreign key was upgraded from an old version, or
will be synthesized using the same algorithm as before if being created
in 19.2-mixed or later.

The other two are "validator" fields that are exact copies of the old
19.1 foreign key representation that a particular foreign key was
upgraded from. These will *also* be synthesized by the creation of new
foreign keys in 19.2-mixed. These fields are used to validate that we
didn't do anything wrong when downgrading a new-style foreign key in the
mixed version state.

---------------------------------------------------------------------

Here is the "state diagram" of how all of the above works:

*19.1*

- everybody writes old-style fks, nobody understands new-style fks.

*19.2-mixed*: a new 19.2 node enters the cluster, but the cluster
              version is not yet flipped.

- 19.2 nodes always write legacy idx fields when synthesizing table
  descriptors.
- 19.2 nodes always assume the existence of legacy idx fields in all
  table descriptors.
- 19.2 nodes unconditionally upgrade all table descriptors from disk
  into the new format, preserving both the legacy idx fields and a special
  "validator" struct that is the entirety of the old-style foreign key.
- 19.2 nodes unconditionally downgrade all in-memory table descriptors
  to the old format, making sure the downgrade was correct by exactly
  matching the result to the special "validator" struct above.
- 19.1 nodes continue on their merry way. No new-style descriptors are
  ever on disk, so they don't know anything is up.

*19.2-final*: all nodes are 19.2, and the upgrade switch is flipped.

- 19.2 nodes behave identically to the above, except:
- 19.2 nodes do not downgrade in-memory table descriptors to the old
  format.

There's a second commit that adds mixed-version tests:

sql: add mixed-version logic tests for fk upgrade

This commit adds a new logic tests configuration that runs logic tests
in the mixed-version cluster configuration that will be visible during a
mixed 19.1/19.2 state. Specifically, this means that the "binary
version" will be 19.2 but the minimum cluster version will be 19.1 (even
though for the purposes of this logic test there will not be any actual
19.1 nodes anywhere).

There are a couple of "known issue" logic tests that are disabled /
hobbled in the fk and schema_change_in_txn logic test configurations,
which necessiated splitting these tests into two nearly identical
configs that are run with and without the mixed version configuration.

Pending the resolution of #39037, the auto-creation of origin indexes
for fks on empty tables doesn't work in mixed-version clusters.

Pending the resolution of #37712, validating a foreign key constraint
within a transaction containing another schema change on the same table
also doesn't work in mixed-version clusters.

Finally, there's also a small difference in the order of constraint
printing for show create table statements in mixed-version clusters.

Co-authored-by: Lucy Zhang <lucy@cockroachlabs.com>
Co-authored-by: Jordan Lewis <jordan@cockroachlabs.com>

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39383)
<!-- Reviewable:end -->

jordanlewis committed:
sql: use new foreign key representation everywhere

This commit does the following salient things:

- change all in-memory foreign key operations to use the new foreign key
  representation, which is defined on table descriptor instead of index
  descriptor.
- change all reads of serialized table descriptor protos to
  unconditionally upgrade themselves from the old descriptor version
  into the new descriptor version.
- change all serialization paths of table descriptors to downgrade
  themselves from the new descriptor version to the old descriptor
  version *when the cluster version is less than 19.2*.

Note that this commit *is not intended to change* any actual foreign key
behavior, such as restrictions on dropping or flexibility in adding.

Note further that this commit puts the state of this branch into peril
with regards to backup and restore on mixed version clusters, which
won't work. This is a known issue and will be resolved before 19.2 is
released. See issue #39474.

There are a couple of other relevant technical details.

New-style foreign keys contain 4 special fields that are designed to
help with this upgrade.

Two are the "pinned" indexes that the foreign key is related to. These
will be the same as the indexes that the foreign keys used to live on in
19.1 and before, if the foreign key was upgraded from an old version, or
will be synthesized using the same algorithm as before if being created
in 19.2-mixed or later.

The other two are "validator" fields that are exact copies of the old
19.1 foreign key representation that a particular foreign key was
upgraded from. These will *also* be synthesized by the creation of new
foreign keys in 19.2-mixed. These fields are used to validate that we
didn't do anything wrong when downgrading a new-style foreign key in the
mixed version state.

---------------------------------------------------------------------

Here is the "state diagram" of how all of the above works:

*19.1*

- everybody writes old-style fks, nobody understands new-style fks.

*19.2-mixed*: a new 19.2 node enters the cluster, but the cluster
              version is not yet flipped.

- 19.2 nodes always write legacy idx fields when synthesizing table
  descriptors.
- 19.2 nodes always assume the existence of legacy idx fields in all
  table descriptors.
- 19.2 nodes unconditionally upgrade all table descriptors from disk
  into the new format, preserving both the legacy idx fields and a special
  "validator" struct that is the entirety of the old-style foreign key.
- 19.2 nodes unconditionally downgrade all in-memory table descriptors
  to the old format, making sure the downgrade was correct by exactly
  matching the result to the special "validator" struct above.
- 19.1 nodes continue on their merry way. No new-style descriptors are
  ever on disk, so they don't know anything is up.

*19.2-final*: all nodes are 19.2, and the upgrade switch is flipped.

- 19.2 nodes behave identically to the above, except:
- 19.2 nodes do not downgrade in-memory table descriptors to the old
  format.

Release note: None

Co-authored-by: Lucy Zhang <lucy@cockroachlabs.com>
Co-authored-by: Jordan Lewis <jordan@cockroachlabs.com>

jordanlewis committed:
sql: add mixed-version logic tests for fk upgrade

This commit adds a new logic tests configuration that runs logic tests
in the mixed-version cluster configuration that will be visible during a
mixed 19.1/19.2 state. Specifically, this means that the "binary
version" will be 19.2 but the minimum cluster version will be 19.1 (even
though for the purposes of this logic test there will not be any actual
19.1 nodes anywhere).

There are a couple of "known issue" logic tests that are disabled /
hobbled in the fk and schema_change_in_txn logic test configurations,
which necessiated splitting these tests into two nearly identical
configs that are run with and without the mixed version configuration.

Pending the resolution of #39037, the auto-creation of origin indexes
for fks on empty tables doesn't work in mixed-version clusters.

Pending the resolution of #37712, validating a foreign key constraint
within a transaction containing another schema change on the same table
also doesn't work in mixed-version clusters.

Finally, there's also a small difference in the order of constraint
printing for show create table statements in mixed-version clusters.

Release note: None

Co-authored-by: Lucy Zhang <lucy@cockroachlabs.com>
Co-authored-by: Jordan Lewis <jordan@cockroachlabs.com>

yuzefovich reviewed (commented):
Wow, there is a lot of work you guys put in here! I left some comments, nits, and questions.

> Reviewed 53 of 53 files at r1, 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-:-LmGRajf0XrxUq4zY-IP:b-75x0s4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @dt, @jordanlewis, and @RaduBerinde)

___
*[pkg/ccl/backupccl/show_test.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFSyCP2YCVhI_43KHY:-LmFSyCP2YCVhI_43KHZ:bigrqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/ccl/backupccl/show_test.go#L199)):*
> ```Go
> 				b INT8 NULL,
> 				CONSTRAINT "primary" PRIMARY KEY (a ASC),
> 				INDEX fkreftable_auto_index_fk_b_ref_fksrc (b ASC),
> ```

[nit]: it might be worth adding a comment here why we're explicitly adding this index.
___
*[pkg/sql/create_table.go, line 598 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGCDBrA-q9-JMOmDfc:-LmGCDBrA-q9-JMOmDfd:bb5iexg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/create_table.go#L598)):*
> ```Go
> 				"foreign key requires an existing index on columns %s", colNames.String())
> 		}
> 		id, err := addIndexForFK(tbl, srcCols, constraintName, ts)
> ```

Should we be adding a new index here regardless of the cluster version?
___
*[pkg/sql/delete.go, line 440 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGDQdB3T_NGvP7F3YZ:-LmGDQdC5kL6UM3jGcVr:b-n93c88) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/delete.go#L440)):*
> ```Go
> 			if _, ok := interleavedIdxs[fk.OriginTableID]; !ok {
> 				// This table is referenced by a foreign key that lives outside of the
> 				// interleaved hierarchy,  so we can't fast path delete.
> ```

[nit]: `s/  so/ so/`.
___
*[pkg/sql/descriptor.go, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGDptZ0elF4Jjac8y2:-LmGDpt_C5CsajlgmQOg:bszyi58) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/descriptor.go#L219)):*
> ```Go
> 		switch t := desc.Union.(type) {
> 		case *sqlbase.Descriptor_Table:
> 			table := desc.GetTable()
> ```

I'm probably missing something, but maybe `desc.GetTable` should be calling `MaybeFillInDescriptor` itself so that table descriptor is always filled in?
___
*[pkg/sql/drop_index.go, line 262 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGFVFZ6WOAk4_kr0B4:-LmGFVFZ6WOAk4_kr0B5:b-g0vovi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d05f199895cb830dd22a679017f2d99b59c8d551/pkg/sql/drop_index.go#L262)):*
> ```Go
> 			// reference on the other table involved in the FK, in case of rollbacks
> 			// (#38733).
> 			idxEntry.ForeignKey = sqlbase.ForeignKeyReference{}
> ```

I'm not sure whether this removal is expected, so I'm leaving this comment to bring attention to it.
___
*[pkg/sql/drop_table.go, line 444 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGHB1E8kjggYjSX7X1:-LmGHB1E8kjggYjSX7X2:b-9rth67) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/drop_table.go#L444)):*
> ```Go
> 	matchIdx := -1
> 	for i, fk := range originTableDesc.OutboundFKs {
> 		if fk.ReferencedTableID != referencedTableDesc.ID {
> ```

[super nit]: it'd be nice to have symmetrical to `removeFKBackReferenceFromTable` loop in here.
___
*[pkg/sql/lease.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGJlNpAI445G3904yO:-LmGJlNpAI445G3904yP:b-320nzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/lease.go#L198)):*
> ```Go
> 		}
> 		if err := tableDesc.MaybeFillInDescriptor(ctx, txn); err != nil {
> 			return nil
> ```

Probably we should be returning `err` here.
___
*[pkg/sql/lease.go, line 1739 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGJyov0YsH2j_wedIg:-LmGJyowDXujyCRB5MBi:b-oa0hkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/lease.go#L1739)):*
> ```Go
> 					case *sqlbase.Descriptor_Table:
> 						table := union.Table
> 						// Note that we don't need to "fill in" the descriptor here. Nobody
> ```

Hm, if we don't need to fill in the descriptor, why not remove this part completely?
___
*[pkg/sql/lease.go, line 1742 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGK6ES1GhKTbA1OwEX:-LmGK6ET7iKVO3Klxyg5:bjzoxaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/lease.go#L1742)):*
> ```Go
> 						// actually reads the table.
> 						if err := table.MaybeFillInDescriptor(ctx, nil); err != nil {
> 							// TODO (lucy): log error
> ```

We could address the TODO now.
___
*[pkg/sql/opt_catalog.go, line 521 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGKSCiCJybn4LXb8Ri:-LmGKSCiCJybn4LXb8Rj:b-bred3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/opt_catalog.go#L521)):*
> ```Go
> 
> func newOptTable(
> 	ctx context.Context,
> ```

I don't see where `ctx` is used within this method.
___
*[pkg/sql/schema_changer.go, line 1500 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGKzSU1VOx6-N3Qzx2:-LmGKzSU1VOx6-N3Qzx3:b-aw7gvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/schema_changer.go#L1500)):*
> ```Go
> 					return err
> 				}
> 				// Get the foreign key backreferences to remove
> ```

[nit]: missing period.
___
*[pkg/sql/schema_changer.go, line 1696 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGLUsi334O3bS5ixKm:-LmGLUsi334O3bS5ixKn:b7cnxy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/schema_changer.go#L1696)):*
> ```Go
> 		for i, fk := range desc.OutboundFKs {
> 			if fk.Name == constraint.ForeignKey.Name {
> 				desc.OutboundFKs = append(desc.OutboundFKs[:i], desc.OutboundFKs[i+1:]...)
> ```

We probably can break out of the loop (and store the fact that we removed the FK).
___
*[pkg/sql/schema_changer.go, line 1699 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGLNal8TKrHvYgjc57:-LmGLNal8TKrHvYgjc58:bdwstfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/schema_changer.go#L1699)):*
> ```Go
> 			}
> 		}
> 		if log.V(2) {
> ```

I think this `if` should only happen when we didn't actually remove a FK from `desc.OutboundFKs`.
___
*[pkg/sql/schema_changer.go, line 2101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGL7ke7-naNTUB2v85:-LmGL7ke7-naNTUB2v86:b-8np5fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/schema_changer.go#L2101)):*
> ```Go
> 						table := union.Table
> 						if err := table.MaybeFillInDescriptor(ctx, nil); err != nil {
> 							return
> ```

Should we log this `err`?
___
*[pkg/sql/scrub_fk.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGLfST-AMB6Fqmxo9H:-LmGLfSU-Np6XU5fpyNw:b-aw7gvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/scrub_fk.go#L110)):*
> ```Go
> 	}
> 
> 	// Get primary key columns not included in the FK
> ```

[nit]: missing period.
___
*[pkg/sql/show_create.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGMHdm7dpGO6EnVw21:-LmGMHdm7dpGO6EnVw22:b-s0g9kg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/show_create.go#L187)):*
> ```Go
> 		f.WriteString(desc.PrimaryKeyString())
> 	}
> 	// TODO (lucy): Possibly include FKs being validated here
> ```

Is this TODO already happening?
___
*[pkg/sql/show_create.go, line 239 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGMRzqBigDZB28Cc4e:-LmGMRzqBigDZB28Cc4f:b-aw7gvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/show_create.go#L239)):*
> ```Go
> 	}
> 
> 	// TODO (lucy): probably go back to using desc.Checks for consistency
> ```

[nit]: missing period.
___
*[pkg/sql/logictest/logic.go, line 420 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGQuvU8rBGwZVKZdG9:-LmGQuvU8rBGwZVKZdGA:bymtrdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/691f62eef46e7caff00c054f7a173d40f1508e09/pkg/sql/logictest/logic.go#L420)):*
> ```Go
> 	},
> 	{
> 		name:                "local-mixed-19.1-19.2",
> ```

Do we care about distributed mixed config?
___
*[pkg/sql/row/fk_existence_delete.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGPJ1d98xN4CUNcsbV:-LmGPJ1d98xN4CUNcsbW:bkb7rb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/row/fk_existence_delete.go#L59)):*
> ```Go
> 			continue
> 		}
> 		//participating := false
> ```

Do we want to keep this commented out code?
___
*[pkg/sql/row/fk_existence_delete.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGPfCqEoKMpuZKRsWf:-LmGPfCqEoKMpuZKRsWg:bkeh31j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/row/fk_existence_delete.go#L85)):*
> ```Go
> 			OriginTableID:       ref.ReferencedTableID,
> 			OriginColumnIDs:     ref.ReferencedColumnIDs,
> 			// N.B.: Back-references always must have SIMPLE match method, because ... TODO(jordan): !!!
> ```

I don't have a good grasp to understand the importance of this comment. Should the TODO be addressed in this PR?
___
*[pkg/sql/sqlbase/structured.go, line 2567 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFWdR13D5BQAhuF8PI:-LmFWdR13D5BQAhuF8PJ:byc0jfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/sqlbase/structured.go#L2567)):*
> ```Go
> 		for i, ref := range desc.OutboundFKs {
> 			if ref.Name == name {
> 				foundIdx = i
> ```

I think we can break from the loop here.
___
*[pkg/sql/sqlbase/structured_test.go, line 655 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFZZKbAjiuYTTbMoOc:-LmFZZKbAjiuYTTbMoOd:bz1n14o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/sqlbase/structured_test.go#L655)):*
> ```Go
> 		},
> 		{
> 			err: `missing fk back reference fk to "foo" from "baz"`,
> ```

This error message "fk back reference fk" seems a little confusing to me. Maybe we want to put the "second" `fk` in quotes?
___
*[pkg/sql/sqlbase/table.go, line 427 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFb7fI1Z5tnF1Rxnkk:-LmFb7fI1Z5tnF1Rxnkl:b-rd34v7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/sqlbase/table.go#L427)):*
> ```Go
> 		return &referencedTable.PrimaryIndex, nil
> 	}
> 	// If the PK doesn't match, find the index corresponding to the referenced column.
> ```

[nit]: `s/column/columns/`.
___
*[pkg/sql/sqlbase/table.go, line 442 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFbXNe2NCiEOy2IBY0:-LmFbXNe2NCiEOy2IBY1:b-hbmgam) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/sqlbase/table.go#L442)):*
> ```Go
> // FindFKOriginIndex finds the first index in the supplied originTable
> // that can satisfy an outgoing foreign key of the supplied column ids.
> // If there's no found index, nil is returned.
> ```

[nit]: for symmetry, I'd either remove the sentence about the case when index is not found here or add such sentence to the comment above `FindFKReferencedIndex`.
___
*[pkg/sql/sqlbase/table.go, line 451 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFbPNO2PiWQQleBLPX:-LmFbPNO2PiWQQleBLPY:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/sqlbase/table.go#L451)):*
> ```Go
> 		return &originTable.PrimaryIndex, nil
> 	}
> 	// If the PK doesn't match, find the index corresponding to the origin column.
> ```

ditto


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
I don't feel like I can provide much feedback on the lower-level details but skimmed through the PR anyway and left some comments. Great work getting this out!

My main general concern is about testing. As I understand it we've currently added mixed-version testing in a configuration where there is one 19.2 node that thinks it's part of a 19.1 cluster, and therefore exercises the downgrade path. Is this sufficient? I'm not sure. I think we should also be testing that a 19.1 node continues to work fine even with a 19.2 node present in the cluster. I guess the assumption is that a 19.2 node will always downgrade descriptors, but I think we should still test this assumption. Backups are another thing I think we should be testing more explicitly, but I understand that we need #39474 in first.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-:-LmK_fMv7hPgQrstgXFs:bgr3paz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jordanlewis, and @RaduBerinde)

___
*[pkg/ccl/backupccl/restore.go, line 70 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmKU7l16zXlgxYFSdS2:-LmKU7l16zXlgxYFSdS3:b-vvd5m1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/ccl/backupccl/restore.go#L70)):*
> ```Go
> 	ctx context.Context, uris []string, settings *cluster.Settings,
> ) ([]BackupDescriptor, error) {
> 	// TODO (lucy): do table descs need to be upgraded here?
> ```

Does this TODO need to be resolved or is it addressed by #39474?
___
*[pkg/ccl/backupccl/show_test.go, line 189 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmKUU5PEEki1z4bX1dJ:-LmKUU5QAazyhNSzscYh:bud1fq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/ccl/backupccl/show_test.go#L189)):*
> ```Go
> 	}
> 
> 	// Foreign keys that were not included in the backup are not mentioned in
> ```

nit: update comment?
___
*[pkg/sql/alter_table.go, line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmKVEdc76u-yZoUZwd-:-LmKVEddEMJWFrG98V6S:b-50crlv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/alter_table.go#L287)):*
> ```Go
> 					}
> 				}
> 				if err := n.tableDesc.Validate(params.ctx, params.p.txn, nil /* clusterVersion */); err != nil {
> ```

The last `nil` parameter doesn't seem to be used. Should we remove it (maybe not in this PR)?
___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmKZ4Ey0jCZtPmG3QfS:-LmKZ4EzAMzKS2rspRll:b2il00m) ([raw file](https://github.com/cockroachdb/cockroach/blob/691f62eef46e7caff00c054f7a173d40f1508e09/pkg/sql/logictest/testdata/logic_test/crdb_internal#L1)):*
> ```
> # LogicTest: local fakedist fakedist-metadata fakedist-disk
> ```

Could you add a comment as to why we're not running under `local-mixed` in the test files that explicitly list the configs?


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
TFTRs!

@asubiotto Yeah, I agree that we need more mixed-version testing with 19.1 nodes. I've done some manual testing by running TPCC (and other FK-related schema changes) on a cluster with both 19.1 and 19.2 nodes, from one of the 19.1 nodes, and didn't discover any issues that way. We should write a roachtest to do that, and do some more intensive testing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-:-LmQ_7Sa6j5aEueF_BH2:bq0ywe0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @dt, @jordanlewis, @RaduBerinde, and @yuzefovich)

___
*[pkg/ccl/backupccl/restore.go, line 70 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmKU7l16zXlgxYFSdS2:-LmQ7Hv08d8zO4BUaqEQ:bg1rohj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/ccl/backupccl/restore.go#L70)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Does this TODO need to be resolved or is it addressed by #39474?
</blockquote></details>

Yeah, it's addressed by #39474. Since this is also documented in an issue and that PR is out, I'll just remove this TODO.
___
*[pkg/ccl/backupccl/show_test.go, line 189 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmKUU5PEEki1z4bX1dJ:-LmQ7daH3YLkdoJhxdn4:bsyzr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/ccl/backupccl/show_test.go#L189)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: update comment?
</blockquote></details>

We're still not including FKs in the output. Previously, if we were omitting the FK, we also omitted the index it was on. (I also replied to Yahor below with more context.)
___
*[pkg/ccl/backupccl/show_test.go, line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFSyCP2YCVhI_43KHY:-LmQ8Yv29fe-6QKo34nZ:bldmx1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/ccl/backupccl/show_test.go#L199)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: it might be worth adding a comment here why we're explicitly adding this index.
</blockquote></details>

This is more of a related fix for a bug/unintended behavior - previously, if we omitted the FK from the output of SHOW CREATE TABLE, we also omitted the index it was on, which now makes even less sense because FKs aren't stored on indexes. I'm not sure a comment here would really be helpful, since it's kind of the "obvious" thing to do. (I suppose it would have been best to make this change in a separate PR.)
___
*[pkg/sql/alter_table.go, line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmKVEdc76u-yZoUZwd-:-LmQA9xMEJX20XmigID5:b-9ikpu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/alter_table.go#L287)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

The last `nil` parameter doesn't seem to be used. Should we remove it (maybe not in this PR)?
</blockquote></details>

Thanks for catching that - I just went ahead and updated `Validate` and `ValidateTable` to remove the extra param.
___
*[pkg/sql/create_table.go, line 598 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGCDBrA-q9-JMOmDfc:-LmQBWCA-eB-S8G4B2AV:b-63b2sg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/create_table.go#L598)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Should we be adding a new index here regardless of the cluster version?
</blockquote></details>

Yeah, we still auto-add indexes in 20.1 because FKs still need an index on the referencing side. Unless you're asking about something else?
___
*[pkg/sql/delete.go, line 440 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGDQdB3T_NGvP7F3YZ:-LmQBxuQ2ooZ9pytqrRk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/delete.go#L440)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: `s/  so/ so/`.
</blockquote></details>

Done.
___
*[pkg/sql/descriptor.go, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGDptZ0elF4Jjac8y2:-LmQCAwG2P0fJIVib5S-:b-fr85tm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/descriptor.go#L219)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I'm probably missing something, but maybe `desc.GetTable` should be calling `MaybeFillInDescriptor` itself so that table descriptor is always filled in?
</blockquote></details>

desc.GetTable is just the auto-generated accessor to get a table descriptor out of the union, so we have to do it afterward.
___
*[pkg/sql/drop_index.go, line 262 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGFVFZ6WOAk4_kr0B4:-LmQCciDAOgXPCM1Y39k:b34uel9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d05f199895cb830dd22a679017f2d99b59c8d551/pkg/sql/drop_index.go#L262)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I'm not sure whether this removal is expected, so I'm leaving this comment to bring attention to it.
</blockquote></details>

Yeah, the functionality and comment were moved to line 151 above.
___
*[pkg/sql/drop_table.go, line 444 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGHB1E8kjggYjSX7X1:-LmQDcM84WZMP8NwQkQZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/drop_table.go#L444)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[super nit]: it'd be nice to have symmetrical to `removeFKBackReferenceFromTable` loop in here.
</blockquote></details>

Done.
___
*[pkg/sql/lease.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGJlNpAI445G3904yO:-LmQDfhtDhmwErt4nx7P:b-m16fak) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/lease.go#L198)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Probably we should be returning `err` here.
</blockquote></details>

Nice catch, thanks.
___
*[pkg/sql/lease.go, line 1739 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGJyov0YsH2j_wedIg:-LmQECOQ66a3GMs69cwH:b-ea644y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/lease.go#L1739)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, if we don't need to fill in the descriptor, why not remove this part completely?
</blockquote></details>

I think we need it so we can call `ValidateTable` below. I updated the comment to be more clear.
___
*[pkg/sql/lease.go, line 1742 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGK6ES1GhKTbA1OwEX:-LmQEAfQ2WoYfyPcTONP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/lease.go#L1742)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

We could address the TODO now.
</blockquote></details>

Done.
___
*[pkg/sql/opt_catalog.go, line 521 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGKSCiCJybn4LXb8Ri:-LmQF5BN0v2Qb8bhOkx_:blhwiuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/opt_catalog.go#L521)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I don't see where `ctx` is used within this method.
</blockquote></details>

Yeah, you're right - this was probably a rebase-related accident or something. I removed it.
___
*[pkg/sql/schema_changer.go, line 1500 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGKzSU1VOx6-N3Qzx2:-LmQFLh-5e9PBmjNmEaT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/schema_changer.go#L1500)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: missing period.
</blockquote></details>

Done.
___
*[pkg/sql/schema_changer.go, line 1696 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGLUsi334O3bS5ixKm:-LmQFNCz6f0XsynEMj16:b-5b20qf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/schema_changer.go#L1696)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

We probably can break out of the loop (and store the fact that we removed the FK).
</blockquote></details>

Yeah, there should have been a `return nil` here. I added it.
___
*[pkg/sql/schema_changer.go, line 1699 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGLNal8TKrHvYgjc57:-LmQFiqz7Bqonc_yazR4:b-bozga1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/schema_changer.go#L1699)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this `if` should only happen when we didn't actually remove a FK from `desc.OutboundFKs`.
</blockquote></details>

I added a `return nil` above.
___
*[pkg/sql/schema_changer.go, line 2101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGL7ke7-naNTUB2v85:-LmQG5UM6jg9wowl3pCI:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/schema_changer.go#L2101)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Should we log this `err`?
</blockquote></details>

Yes.
___
*[pkg/sql/scrub_fk.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGLfST-AMB6Fqmxo9H:-LmQGCGf4Znuhs0E839c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/scrub_fk.go#L110)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: missing period.
</blockquote></details>

Done.
___
*[pkg/sql/show_create.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGMHdm7dpGO6EnVw21:-LmQGGUnEmS9YK7g0s4h:b-mkmq1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/show_create.go#L187)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Is this TODO already happening?
</blockquote></details>

Sort of - `OutboundFKs` includes FKs being validated but not FKs queued in the mutations list to be validated, and I meant the latter. I'll improve this comment.
___
*[pkg/sql/show_create.go, line 239 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGMRzqBigDZB28Cc4e:-LmQGyfu4omztPAeb-mZ:b5zxg6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/show_create.go#L239)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: missing period.
</blockquote></details>

I got rid of this and folded it into the above TODO.
___
*[pkg/sql/logictest/logic.go, line 420 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGQuvU8rBGwZVKZdG9:-LmQH2DLEYlJ6laLmHke:ba87fuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/691f62eef46e7caff00c054f7a173d40f1508e09/pkg/sql/logictest/logic.go#L420)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Do we care about distributed mixed config?
</blockquote></details>

We thought the benefits wouldn't outweigh the complexity of an additional config - as Alfonso said, the real benefit would from mixed-version testing where we have actual 19.1 nodes, which we can't do in a logic test.
___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmKZ4Ey0jCZtPmG3QfS:-LmQZzwd9Qd1MQCyvzw_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/691f62eef46e7caff00c054f7a173d40f1508e09/pkg/sql/logictest/testdata/logic_test/crdb_internal#L1)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Could you add a comment as to why we're not running under `local-mixed` in the test files that explicitly list the configs?
</blockquote></details>

Done.
___
*[pkg/sql/row/fk_existence_delete.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGPJ1d98xN4CUNcsbV:-LmQLnAsDcmbBdEQ6ZAt:b-ynwl21) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/row/fk_existence_delete.go#L59)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Do we want to keep this commented out code?
</blockquote></details>

I removed it.
___
*[pkg/sql/row/fk_existence_delete.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmGPfCqEoKMpuZKRsWf:-LmQM2BK2lXudeSfVmpk:b-aavfo1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/row/fk_existence_delete.go#L85)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I don't have a good grasp to understand the importance of this comment. Should the TODO be addressed in this PR?
</blockquote></details>

I don't think there's anything that can be done. @jordanlewis knows more about this than I do, but this is necessary to do the FK deletion lookups for now. All this code will be removed when the HP is removed in 20.1.
___
*[pkg/sql/sqlbase/structured.go, line 2567 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFWdR13D5BQAhuF8PI:-LmQMkX-2eLkCzK5fDDi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/sqlbase/structured.go#L2567)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think we can break from the loop here.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured_test.go, line 655 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFZZKbAjiuYTTbMoOc:-LmQMykACxxL9mwQXh_g:bu5zla0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/sqlbase/structured_test.go#L655)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This error message "fk back reference fk" seems a little confusing to me. Maybe we want to put the "second" `fk` in quotes?
</blockquote></details>

I think this is just a typo. I removed the second `FK`.
___
*[pkg/sql/sqlbase/table.go, line 427 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFb7fI1Z5tnF1Rxnkk:-LmQN6Zn-6nIAzGwnHis:bgftfit) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/sqlbase/table.go#L427)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: `s/column/columns/`.
</blockquote></details>

Did you mean to leave a comment here?
___
*[pkg/sql/sqlbase/table.go, line 451 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39383#-LmFbPNO2PiWQQleBLPX:-LmQNswX3u66D0a1jYR3:b58lciu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9c8320b40d43ec26e697cef87a541020ca0242f/pkg/sql/sqlbase/table.go#L451)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto
</blockquote></details>

This comment looks the same as the one in the above function - am I missing something?


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
Relaying an offline convo here:

I read/skimmed parts of this on and off over last couple days and they certainly look like they're doing what the description says, and what the description says it does sounds good to me. However I'm not all that familiar with a fair amount of the code changed here, so I can't say much more than that with any real confidence. 

That said, I'm 👍 on getting this merged -- even if there are remaining areas with planned work / tests / etc (Lucy has mentioned backup/restore as one in offline), just since if we think we're doing this for 19.1, the sooner it gets into master, the sooner it starts getting exercised and any issues have a better chance of being found and fixed before release (and less rebase overhead means more time for finding and fixing any real bugs).

lucy-zhang commented:
TFTRs! I'm merging this now, in line with dt's comment above. The remaining work is being tracked in #39504.

bors r+
